### DISCUSSION TITLE
Understanding Retrieval_Time

### DISCUSSION THREAD

**Discussion by matiasandina** [KravitzLabDevices/FED3#2] [Q&A]
I am trying to further understand the way `Retrieval_Time` is calculated and logged to file.
If I understand correctly, these two snippets are crucial to calculating the time of retrieval of a pellet.

```
//get time of pellet removal in unixtime seconds since 1970, and subtract pelletTime to determine 
// how long the pellet was in the well
DateTime now = rtc.now(); 
retInterval = now.unixtime() - pelletTime; 
if (FEDmode == 0 or FEDmode == 11) { 
   // adjust retInterval for free/timed feeding mode delay
   retInterval = retInterval - 3;
   // ...
}
```

```
if (digitalRead(PELLET_WELL) == LOW ) {
  //get time of pellet drop in unixtime seconds since 1970
  DateTime now = rtc.now();
  pelletTime = now.unixtime();
  PelletAvailable = true;
  // ...

}
```

I would like to clarify if my understanding is correct here. If I understand the code correctly, the logged `Retrieval_Time` would be the time difference between a pellet was available in the well and the pellet is removed from the well. If so, the timestamps don't have real information about the retrieval time, they only show when a pellet was removed. However, these timestamps are as close as we can get to the actual moment of eating, is this correct? (I'm banging my head with the interpretation of retrieval time and the actual eating moment, and what would be my variable of interest for each experiment)

For example, I have below a part of a dataset from FED3. `Retrieval_Time` and `MM:DD:YYYY hh:mm:ss` come directly from FED3 logging, `datetime` is the POSIXct conversion of the datetime and `r_time` is the retrieval time calculated by me using a point to point difference of the timestamps.

ID<chr> | MM:DD:YYYY hh:mm:ss<chr> | datetime<S3: POSIXct> | Retrieval_Time<dbl> | r_time<time>
-- | -- | -- | -- | --
HC031 | 12/9/2020 22:12:09 | 2020-12-09 22:12:09 | NA | 323 secs
HC031 | 12/9/2020 22:15:58 | 2020-12-09 22:15:58 | NA | 229 secs
HC031 | 12/9/2020 22:16:22 | 2020-12-09 22:16:22 | 19 | 24 secs
HC031 | 12/9/2020 22:17:30 | 2020-12-09 22:17:30 | 65 | 68 secs
HC031 | 12/9/2020 22:30:23 | 2020-12-09 22:30:23 | 41 | 773 secs
HC031 | 12/9/2020 22:30:58 | 2020-12-09 22:30:58 | 30 | 35 secs
HC031 | 12/9/2020 22:31:22 | 2020-12-09 22:31:22 | 21 | 24 secs
HC031 | 12/9/2020 22:31:48 | 2020-12-09 22:31:48 | 19 | 26 secs
HC031 | 12/9/2020 22:43:47 | 2020-12-09 22:43:47 | 31 | 719 secs
HC031 | 12/9/2020 22:44:12 | 2020-12-09 22:44:12 | 21 | 25 secs

I understand `NA` values come from the 120 seconds threshold on the logger that are stored as "timed out" (`if (retInterval >= 120) logfile.println("Timed_out");` in `Logdata.ino`).  Is there a reason why we would not log the actual value ?
I guess the 3 seconds differences seen on some entries come from adjustment because of delay on free-feeding program, and that matches the differences in some rows (e.g, 65 vs 68 seconds). I am confused about the results for `Retrieval_Time` in cases where the time between pellets is big (e.g, 773 seconds) but FED seems to log 41 seconds. Where is this difference coming from? Looking at the data, it's not explained by motor turns either, this particular case had 1 motor turn and cases with `Motor_Turns` > 1 are often just 3 to 5 seconds apart.

**Comment by KravitzLab** [comment]
Hey Matias!  Thanks for using the discussion!  Sorry for the delay in responding.

Yup the retrieval time is timing the duration from the pellet first being detected and the pellet first being removed, based on reading the pellet well itself.  It's not the moment of actual eating because they could remove the pellet and wait 2 seconds before eating it, or even drop it on the floor, but it's the only information FED3 has about the pellet.  I do know some labs are video-taping the mice from the bottom so they can confirm the mice eat each pellet, but this is because they are trying to understand what a neural population is doing during feeding and therefore want to be certain of the timing of eating.  But that's a ton of work, so I think retrieval time can be used as a proxy for learning or motivation.  For instance, when we train mice over their first experience with FR1, we find the retrieval time drops from ~10s on average to ~5seconds in the first night, demonstrating the quickening of pellet retrieval behavior after a successful poke (see Fig5 here: https://www.biorxiv.org/content/10.1101/2020.12.07.408864v1).  I'm not entirely certain what this means in terms of motivation, but I think it's evidence of them learning to use FED3.

To answer your questions, no there's not a reason we can't log the retrieval times >120s, I think I made a decision at some point that they weren't really meaningful, what does it mean if the mouse pokes and doesn't grab the pellet?  I think it means he poked because he was exploring but either doesn't want the pellet or doesn't understand the link between the poke and the pellet.  At any rate I probably shouldn't make this decision for other people, so I can change it to log the real retrieval time.  I'll update that in the newer code.  Also, in our FED3VIZ analysis program you can select different thresholds for this cutoff.  I think a cutoff is important though, otherwise you will end up with some "retrieval time" values >1 hour, which don't make sense to average in.

For your other question, I'm not sure where r1 is coming from.  How are you calculating that?  FED3 doesn't write the pellet detection time to the SD card, so you can't calculate it yourself from the poke timestamps.  Are you subtracting the time from the previous pellet?  

It's also possible you found a bug in this calculation. Could you attach a snippet of the original data file highlighting the discrepancy you're seeing?  Thanks!

**Comment by matiasandina** [comment]
Hi,
There's definitely a need of a cutoff. I guess keeping it as a double can help keep the `Retrieval_Time` column without `NAs` (converted from "timed out"). The user has to filter out the `NAs` or the high values. Either way there's a post-processing needed. I can see why choosing 120 seconds and adding something that is not a number can force the user to be closer to the data instead of just calculating the mean of a whole column that is long tailed distributed.

I'm currently using free feeding mostly, so for me, there's always a pellet on the well and there's no poke-eat sequence that we are looking into. I was trying to see whether there was any relationship between the `Retrieval_Time` and the time between pellets (as measured by the time elapsed between the timestamps on successive pellet events). I should probably not call both things retrieval time because it leads to confusion. The retrieval time is what the FED3 calculates. Using the same terminology that you used in the paper, what I'm calculating is the inter-pellet time. To calculate the time between eating, I calculate the point to point derivative of the timestamp of each pellet event. I assume that if animals removed a pellet, they are actually eating.

If `Retrieval_time` is under 120s, I should expect a linear relationship between retrieval time and inter-pellet time. Small jams and failure to deliver a pellet for any reason would add a little bit more time to the time between pellets, but the relationship should be there. Above 120s, I would expect `NA` for `Retrieval_Time` and whatever number of seconds between each pellet was removed and logged to text. 

What was strange for me was the events with very high inter-pellet time and very low `Retrieval_Time`, it seemed to me at first that maybe the motor was jamming for a long time (and thus no pellet could be eating, with the mouse waiting at the well to eat it as soon as it was delivered and hence show low retrieval time). But the `Motor_Turns` are low for these events.
 
I should probably have posted the data in the previous post, but anyway here are the series of plots/data that got me confused on the first place. I renamed `r_time` to `eating_time` although it's actually the inter-pellet interval.

This is a plot of a cohort of animals running on free-feeding, `Retrieval_Time` as given by FED3 vs interpellet-interval as calculated by the point to point difference in pellet timestamps. Here, it's already evident that for a good number of points, `Retrieval_Time` < 120 can happen while interpellet-time is way greater than 120 seconds. 

![image](https://user-images.githubusercontent.com/7494967/103091139-2f547180-45c1-11eb-873d-4812c01bb1a9.png)

From here, I went to look for my expected linear relationship plus noise given by the motor turns. Indeed, the following plot shows that. For `Motor_Turns = 1` the relationship is great, except when it's not!  

![image](https://user-images.githubusercontent.com/7494967/103091009-cb31ad80-45c0-11eb-87e1-4cc661dfed1e.png)

I have been thinking about what hardware problems can be compatible with the data. What can trick the retrieval time to start over even though the animal did not take a pellet? 

* If, for example, a pellet is removed from the path of the photointerrupter but it very quickly comes back (maybe because the animal failed to retrieve it but moved enough?), the motor would not actually dispense a new one and the event would not be logged to the SD.  Would this make sense? Can it happen multiple times as the animal is trying to get the pellet so each time the reset happens we are still under 120 seconds but actually no pellet was registered as eaten?

* Is there any combination of a FED delivering more than one pellet that can trick this? 

I can always be wrong in my math so feel free to check the raw data for one animal. 


[FED003_120920_00.txt](https://github.com/KravitzLabDevices/FED3/files/5739798/FED003_120920_00.txt)

**Reply by KravitzLab** [reply]
Thanks for the analysis!  Yes I think you're pushing against the limits of the hardware here, and identifying an edge case where the mouse is interacting with the FED in a way that it is not robust against.  Do you know what version of the code is on your FED3 BTW?

One possibility that could be it: I've seen a device get stuck asleep without a pellet in the well.  To give you some background on the sleep functionality, the FED3 checks for a pellet in the well and whenever it sees one it tries to go to sleep to save battery life.  However, if it checks and sees a pellet and decides to go to sleep, AND the mouse grabs the pellet right after this sleep command is given, it is possible for the FED3 to sleep without a pellet being in the well.  In this case, your eating_time clock would keep running but there is no pellet for him to eat so the retrieval time clock would not start.  So FED could be just sitting there for a couple minutes with no pellet, and when it wakes up (which would be caused by the mouse poking around with the photointerrupters) it realizes there's no pellet, it drops one, and the mouse takes it?  I think that could cause what you're seeing?  It's a very hard issue to replicate, unless you have the ability to record continuous video of the pellet well and go back to specific pellet times to confirm if that's what's happening.  Can you calculate how often this is happening? 

I have been mostly revising the library version of the FED3 code these days and I've stopped this from happening in that version because I'm using a different sleep library that allows the FED to wake up once per second and check the photointerrupters to protect against this specific problem. Would you be willing to testing out a version of the free-feeding program that uses the FED3 library and seeing if the issue goes away?  That might be the fastest way to nail this down?

**Comment by matiasandina** [comment]
Right now, I am using Version 1.1.44. What you say about sleep time makes sense, not sure how the wake up works in free feeding mode, but this functionality seems like something that we actually want to be built-in, to save battery while the animals are not taking pellets or interacting with the device. 
For me this is not a deal-breaker, I was just checking things made sense (for the most part they do). I would mostly ignore the retrieval time in free-feeding mode and just care about the inter-pellet interval, but I am happy to test the FED library if that can help parse out the issue. I would need to run [this](https://github.com/KravitzLabDevices/FED3_library/blob/main/examples/FreeFeeding/FreeFeeding.ino) example right?

**Reply by KravitzLab** [reply]
Cool, if you calculate how often this is happening you may decide it's fine to filter out these buggy events with the logic that you couldn't get an accurate retrieval time on all trials, but it still could be useful to look at retrieval time on those trials where it is accurate. 

Without sleep FED3 will last ~1-2 days on the battery, with sleep it lasts ~7-10! (Exact battery life depend on the program that's running and how often the mouse triggers the motor).  So it is pretty important to keep sleep functionality.  But it can create bugs if it goes to sleep in the wrong state due to weird/unpredictable rodent behavior.

At any rate, the FED3 library uses a different sleep library than the base code (The FED3 library uses [ArduinoLowPower ](https://www.arduino.cc/en/Reference/ArduinoLowPower) for sleep.  I'd like to change this in the base code too but haven't found the time to do it yet).  ArduinoLowPower has the microcontroller wake up every second and check the status of all of the inputs.  If everything is good it goes back to sleep, otherwise it stays awake until it encounters the sleep command again.  This checking takes a very short period of time (<1ms I think but I haven't timed it) so it doesn't drain the battery like keeping everything awake, but it is a safeguard against it getting stuck in a strange state.  I'd love to see your correlations after running a mouse with the library version to see if it fixes this.

Thanks for testing out the library.  For almost everything it uses the exact same code as the base 1.4.44 code so it shouldn't change anything, but it is MUCH easier to use if you want to customize the function of FED3.  You can check out the examples to see how easy it is to write custom programs.  And yes that's the example for free feeding!  The easiest way to use it is to install v1.0.5 of the FED3 library via the library manager in the Arduino IDE: 
![image](https://user-images.githubusercontent.com/16770585/103174486-34ffc080-4828-11eb-8a6e-bbef524d6275.png)

And then just click on that example: 

![image](https://user-images.githubusercontent.com/16770585/103174773-5feb1400-482a-11eb-9821-af7e46949647.png)

Feedback/bug reports/no-bug reports welcome., thanks again! -Lex

**Comment by matiasandina** [comment]
Hello again, 

sorry for the delay, we were running a cohort and I didn't want to switch libraries in the middle of it. I am using the library with 3 seconds timeout on FreeFeeding mode. I also included here 2 FEDs that were using the previous version.

A few things I notice are, 
1) The new library logs all motor turns as 0 ?
2) It looks like the new library takes 3 seconds timeout and THEN tries to dispense a pellet. I think the old version somehow is faster than that (closer to the line with lope 3 seconds). I can sense this while trying the FEDs myself during calibration.

I see some weird errors on the new library too, `Retrieval_Time` close to zero but actual difference between timestamps in the thousands

![image](https://user-images.githubusercontent.com/7494967/106066332-a30b0200-60ca-11eb-9c4a-8fbdb28bd435.png)

Here is the 120 seconds range, also most of the errors happen weirdly with `Retrieval_Time` close to zero

![image](https://user-images.githubusercontent.com/7494967/106065563-799da680-60c9-11eb-8045-f514320fd7c8.png)

I'm not sure the new library is recording retrieval time better than the old one.

Here are some tables with examples

```
  Device_Number LibaryVersion eating_time Retrieval_Time
          <dbl> <chr>         <drtn>               <dbl>
1             5 NA             122 secs           118   
2             7 1.1.0          132 secs             0   
3             7 1.1.0          844 secs             0.06
4             9 1.1.0          129 secs             0.06
5             9 1.1.0         2439 secs             0.2 
6             9 1.1.0          163 secs             0.06
7            12 1.1.0         2103 secs             0.4 
```

And here is a histogram of the differences between `eating_time` and `Retrieval_Time` (only including `eating_time` < 120 here)

![image](https://user-images.githubusercontent.com/7494967/106067036-d13d1180-60cb-11eb-857d-dd69cf8af823.png)
![image](https://user-images.githubusercontent.com/7494967/106067268-53c5d100-60cc-11eb-9a43-04a554b98ea5.png)

 
Finally, this is a separate comment. The column `LibaryVersion` has a typo (and also does not follow convention of `Upper_Case` like the other variables).

**Comment by KravitzLab** [comment]
Thanks for documenting this Matias!  I opened "issues" for your comments on LibraryVersion and motor turn counting.  I've fixed the Library_version typo and confirmed the bug you found with motor counts all being written as "0".  I'll look into that bug this week.  Thanks for pointing these out!  I've also made some additional bug fixes and am now on version 1.2.3, can you update to that for additional testing?

I think the timeout you noticed may be a "feature" :)  I set a default timeout of 10s in the example FreeFeeding script, but you can disable it by removing line 21 (if you don't set a timeout it will default to 0).  Can you check if this is the cause of what you're seeing?
![image](https://user-images.githubusercontent.com/16770585/106620812-8db52e00-6537-11eb-95a7-3b264d36d77a.png)
The desired behavior of this program is to drop a pellet, wait until it is removed, start and complete the timeout, and then drop the next pellet.  Is that what you're seeing?  Rather than respond here do you want to move this to an "issue" so we can keep it straight from your other question?

Your last issue is deeper.  I wonder if the FED3 is somehow "missing" the pellet removal which may account for the huge time delays between the next pellet.  Can you try downloading the latest library 1.2.3 and seeing if this is fixed?  I made some adjustments to the dispensing function since 1.1.1.  Also, how old are your 3D prints?  We made some edits to the design in September to try to block off the pellet area more and inhibit mice from taking two pellets.  The issue we found was that they could grab a pellet so quickly that the FED wouldn't detect it and would dispense a second one.  Here is a photo of the [newest design]( https://www.tinkercad.com/things/0QaiVw7KR3Y) that has the two rectangular guards above the pellet well to prohibit pellet stealing. Does yours look like this?
![image](https://user-images.githubusercontent.com/16770585/106628549-85f98780-653f-11eb-83de-c0a59be500ae.png)

**Reply by matiasandina** [reply]
Sure, I will try to download the latest version to run some trials.
Prints are the first version of the angled.
We just put an order for 20 more FEDs so we might be in for some printing and I can give a try to this new version. This version will also probably prevent pellets bouncing sideways out of the well.
Because the animals live with the FEDs 24 hs in my case, I don't see them stealing pellets (meaning I don't stay there long enough). 4-6 months adult mice (30-35 gr) consume around 200-300 20 mg grain pellets per day according to the log, so I'm not sure they would actually eat much more.

I'm not sure I understand how this is connected to `eating_time` being different from `Retrieval_Time`
